src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs

```diff
@@ -644,26 +646,41 @@ private void Dispose()
                 {
                     if (!_noImplicitTransforms)
                     {
+                        // given the type, fetch the implicit transforms for that type and put it in the implicitTransforms variable.  
                         Type argType = args.GetType();
-                        if (_expectedArgType != argType)
+                        TransformSpec implicitTransforms;
+
+                        // First check the one-element cache _firstImplicitTransformsEntry
+                        ImplicitTransformEntry cacheEntry = _firstImplicitTransformsEntry;
+                        if (cacheEntry != null && cacheEntry.Type == argType)
+                        {
+                            implicitTransforms = cacheEntry.Transforms;     // Yeah we hit the cache.  
+                        }
+                        else if (cacheEntry == null)
                         {
-                            // Figure out the default properties to send on to EventSource.  These are all string or primitive properties.  
-                            _implicitTransforms = null;
-                            TransformSpec newSerializableArgs = null;
-                            TypeInfo curTypeInfo = argType.GetTypeInfo();
-                            foreach (var property in curTypeInfo.DeclaredProperties)
+                            // _firstImplicitTransformsEntry, we should fill it.  
+                            implicitTransforms = MakeImplicitTransforms(argType);
+                            Interlocked.CompareExchange(ref _firstImplicitTransformsEntry, 
```

Comment:
If we lose the race here, don't we need to update implicitTransforms?  Otherwise this thread is pointing to the wrong one.  Seems like this needs to be something more like:
```C#
Interlocked.CompareExchange(ref _firstImplicitTransformsEntry, new ImplicitTransformEntry() { Type = argType, Transforms = implicitTransforms }, null);
implicitTransforms = _firstImplicitTransformsEntry;
```
?

Reply:
The cache is purely an optimization in the sense that each thread could have called MakeImplictTransforms each time it logged an event and it would have been OK.   Thus that fact that when you lose the race you are not using the same object as other threads is OK.   

That being said I was going to add your suggestion, but it would need to fixed to check that the _firstImplictTransformsEntry.Type was argType (since we may be racing with a thread with a different arg-type.  

Thus it we need if-check and set, which for code that is not needed, I would rather not do.  

Instead I have simply commented why the current code is OK.   If you really want that additional code however, let me know and I will add it (ultimately this is not an important code path).